### DISCUSSION TITLE
Everything related to new beacons bandits is broken + broke existing beacon mods

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#14291] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

# **The new beacons are not just broken in several ways, they also broke beacon mods, and especially beacon bandits**

Eventset requiring a bandit layer. 
**None of the existing mods have this layer so bandits don't spawn since this wasnt communicated with the update. 
It also means "Bandits" eventset competes with "Encounters", even when the "bandit beacons" don't have encounter spawnpoints.** 
THE LAYER CHECK ITSELF is a problem.
![image](https://github.com/user-attachments/assets/58124f89-e6e0-4996-bfaa-461d1c7f60fe)

Half of the walls are set as indestructible. I first thought of a bug, but no some (not all) interior walls have been manually assigned invulnerable at random while the thick outer wall can be broken.
**(IT SHOULD BE EITHER ALL OF THEM OR NONE OF THEM)**
![image](https://github.com/user-attachments/assets/c62544f5-5022-4ae3-a2a9-68ca26511074)

Some beacons don't have closing circuits for the airlock so they just get flooded by diving inside, which kills the bandits
![image](https://github.com/user-attachments/assets/0b0bd343-6285-4b47-a119-4df6e43b43c7)

NPCs failing to man the invisible "defendable" item (instead of periscopes??).
![image](https://github.com/user-attachments/assets/29eb98c8-3193-4de6-9021-78b598ad93fb)
![image](https://github.com/user-attachments/assets/282f0026-d930-4add-b0db-a44b5f236796)

DIving suits spawning in the same locker and falling outside because the event tags the same locker and spawns all the suits there *(also someone nutted on the suit locker)*
![image](https://github.com/user-attachments/assets/066e2942-f225-4349-ad9c-f28581e69c3e)

NPCs AI being plain stupid and failing to do anything in general, not repairing machines
![image](https://github.com/user-attachments/assets/f74cae6a-1a87-4e6b-93a6-748464c437a2)

An obtuse new system relying on layer functionalities and hidden niche items.
On paper that's interesting to have bandit difficulty scaling and *might* save some time by only editing a single beacon file but in practice unless you're balls deep into modding there is NO way to figure this out, and is prone to more problems than it solves
![image](https://github.com/user-attachments/assets/35e135ab-ab7e-4ff6-af8c-4ce96fe858b4)

# Solutions

* Create NPC prefabs for beacon bandits, equipped with disposable suits instead of spawning suits. They will put the suits in the lockers by themselves. Or just prefill the lockers with disposable suits.

* Set walls as Invulnerable using events (is it possible to tags all the walls efficiently?), **if not possible don't.**

* Try  OrderOption="operateweapons" in NPCOperateAction (I havent tried it), or directly tag the periscopes as operate targets. 

* Get rid of "defendable position" entirely, nobody knows this exists. Increase NPC hearing or make them patrol in NPCprefab if necessary.

* Find a way to not use the `requiredlayer="bandit"` check. Example: trigger the bandit event regardless, have a tag action, if a tagaction doesnt find a bandit spawnpoint it just ends the event. 

### Reproduction steps

Generate a level with BeaconStation_KopraKoht

Also try Beacons Extended
https://steamcommunity.com/sharedfiles/filedetails/?id=2389600483

### Version

v1.5.9.1 (Summer Update Hotfix 2)

**Comment by ubertpendragon** [comment]
PS I don't hate the new layer system in theory but It's a case "too late" and as a result there's no compatibility with existing workshop content.

Getting bandits to work just ends up being too complicated as you HAVE to make a beacon that is both compatible with encounters and bandits at once, instead of just bandits

**Comment by Crazyjr20** [comment]
New system also ignores the submarine editor saving options for Allowing damage types.

**Comment by Regalis11** [comment]
Good points, I think we do need to look into many of these issues. However, I don't entirely agree with all of the points:

> Get rid of "defendable position" entirely, nobody knows this exists. Increase NPC hearing or make them patrol in NPCprefab if necessary.

You don't need to know it exists: the bandits work exactly the way they used to if a beacon doesn't have those. But if you want to make a bandit stay at a specific position (e.g. to watch the entrance or some other strategic position), that can be done with the new "defendable position" item. Increasing their hearing or making them patrol would not allow them to figure out where the designer of the beacon station intended those positions to be.

> Half of the walls are set as indestructible

Afaik this only applies to specific walls - it's not random, I believe just the walls around "safe rooms" are marked as indestructible, pretty much like they were in the old beacon stations.

> Getting bandits to work just ends up being too complicated as you HAVE to make a beacon that is both compatible with encounters and bandits at once, instead of just bandits

Could you elaborate on this, why do they need to be compatible with both? Unless I'm missing something, the "encounter" events should just get ignored if the beacon station only has bandits.

**Reply by Crazyjr20** [reply]
If I can bring up one issue with the new system, its far more limiting in how you can edit stations to seem "bandit inhabited". The layers available only allow us to add new stuff, can't move or remove stuff which can really add to making it seem inhabited. For example I've attached two screenshots of before and after from my own bandit beacon which no longer works and would need to be changed to be more static for the new system. It's not the end of the world, but is a downgrade in potentional from the old system with the main benefit being smaller file size which is a valid benefit. I will admit I haven't properly used the layer system yet, but I'd assume if there was a layer defined for this event specifically and it gets disabled when making it a bandit beacon, then you could remove stuff by adding it to that layer, so that could be one solution.
![before](https://github.com/user-attachments/assets/1d91a614-228e-4913-b98b-6828cd6266fa)
![after](https://github.com/user-attachments/assets/54c40c02-52ca-4073-9f3d-f2e87eb47446)

**Reply by ubertpendragon** [reply]
> > Get rid of "defendable position" entirely, nobody knows this exists. Increase NPC hearing or make them patrol in NPCprefab if necessary.
> 
> You don't need to know it exists: the bandits work exactly the way they used to if a beacon doesn't have those. But if you want to make a bandit stay at a specific position (e.g. to watch the entrance or some other strategic position), that can be done with the new "defendable position" item. Increasing their hearing or making them patrol would not allow them to figure out where the designer of the beacon station intended those positions to be.

I think the issue here is that they're basically an obscure placeholder item combined with an event for something that is basically "hold position" in a certain spot. The average beacon builder wouldnt know

--------

> > Half of the walls are set as indestructible
> 
> Afaik this only applies to specific walls - it's not random, I believe just the walls around "safe rooms" are marked as indestructible, pretty much like they were in the old beacon stations.

The beacon in the screenshot (Koprakoht) has a bunch of invulnerable walls not connected to the saferoom, a few other beacons have this problem too. It's just not intuitive.

If the justification for invincible saferoom that walls need to be invincible to protect special event NPCs, from monsters then **instead you should disable the `beaconstation_populate` eventsets in case one of the beacon-specific events are at play:**

- shockjock2 (unused?)
- prophetofsierpinsk (the end game beacon, no eventset spawns monsters in that region)
- psychoclown (part of encounters, doesnt spawn along monsters)
- traitor mudraptor egg (doesnt really care about other events)

In its current state, shockjock2 doesnt prevent monsters spawning or walls being damaged, or bandits from spawning (bandits killed the NPC as he walked out of saferoom)
![image](https://github.com/user-attachments/assets/1a6c3555-8ff9-4d37-8ca2-90ce0f06fcee)


Easiest way would be for the eventcheck if there's a single character in the beacon before triggering?
I'm not sure how. There's the question of eventset vs scriptedevent load order before the check, the fact that the eventset doesnt even check if there's a beacon itself...

------

> > Getting bandits to work just ends up being too complicated as you HAVE to make a beacon that is both compatible with encounters and bandits at once, instead of just bandits
> 
> Could you elaborate on this, why do they need to be compatible with both? Unless I'm missing something, the "encounter" events should just get ignored if the beacon station only has bandits.

`beaconstation_populate` rolls a dice between:

- Encounters (monsters + damage beacon)
- Bandits
- Nest (crawlers/mudraptors + eggs)

Bandits and Nest use `requiredlayer` to trigger

If your bandit beacon has a bandit layer but no nest layer, then the eventset picks between encounters or bandits.
It's a toss up between encounters (does nothing if no encounter spawnpoint) or bandits.

The old event code used 1 eventset for encounters, and 1 eventset for bandits (with mudraptor/crawler sprinkled), both always triggered and if the spawnpoint wasnt present it wouldnt spawn anything.
**That's this old system the entirety of the workshop content is designed around**

**Reply by Regalis11** [reply]
>  I think the issue here is that they're basically an obscure placeholder item combined with an event for something that is basically "hold position" in a certain spot. The average beacon builder wouldnt know

Yeah, I agree to some extent, but I would also argue that you don't even need to know about the event or how it works "under the hood" - you can take a look at the vanilla beacon station, notice that there's this "placeholder item" in the spots the bandits stay at, and look at the description of that placeholder item to see what it is or what it does.

> If I can bring up one issue with the new system, its far more limiting in how you can edit stations to seem "bandit inhabited". The layers available only allow us to add new stuff, can't move or remove stuff which can really add to making it seem inhabited.

Hmm, I'm not sure I understand why. It's still entirely possible to design the entire station to be exclusively "bandit inhabited", with stuff removed or moved from the "normal" version of the version of the station. The only limitation (which in hindsight, was a bad decision as far as backwards compatibility is concerned), is that the bandit events can only trigger if there's a bandit layer in the station.

> downgrade in potentional from the old system with the main benefit being smaller file size which is a valid benefit.

The file size isn't really a problem in practice (the submarine files are pretty small anyway), nor why we decided to use the layer system: the point was to make it easier to add variety to the stations without having to maintain multiple copies of each station (e.g. a "generic" one, a monster-infested one, a version inhabited by bandits, and a more difficult bandit version with turrets and such...). However, I do agree that we should've taken backwards compatibility into account better.

I would also like to point out that now the vanilla events have been set up so that bandits aren't guaranteed to appear in bandit stations - the likelihood of running into bandits increases with the difficulty. Since that's not how it used to work, the behavior of mods that use those events is ofc different now too. I think if you'd prefer bandits to always spawn in bandit outposts, it could be worth considering to add a custom event that spawns some bandits (and doesn't have the bandit layer requirement, if you don't want to use it), and make that event always trigger, the same way as the old event did.

**Reply by ubertpendragon** [reply]
As a quick summary I will say:

1. Bandit layer would've worked had it been released 2 years ago; too much workshop content does NOT use the system. As it stands all bandit beacons are broken (chance to not spawn bandits). 

2. Need a simple way for "beacon events" like Shockjock2 to disable eventsets to prevent enemies from spawning in the beacon. _DisableEventSetAction_ ? I have no idea

3. Invincible walls need to go

4. The bugs

**Reply by ubertpendragon** [reply]
> the vanilla events have been set up so that bandits aren't guaranteed to appear in bandit stations - the likelihood of running into bandits increases with the difficulty. Since that's not how it used to work, the behavior of mods that use those events is ofc different now too. I think if you'd prefer bandits to always spawn in bandit outposts, it could be worth considering to add a custom event that spawns some bandits (and doesn't have the bandit layer requirement, if you don't want to use it), and make that event always trigger, the same way as the old event did.

I personally could in Beacons Extended, other mods would still be broken unless updated until I go and contact all of them. Then we'll have 5 different eventsets for spawning bandits each.

This would also break vanilla beacons (either by spawning twice as many bandits or bandits that instantly die to flooding/monsters).

**Comment by ubertpendragon** [comment]
I think the simplest possible fix regarding compatibility would be to check for `requiredspawnpointtag` tags instead of `requiredlayer`
All of the existing workshop beacons have proper spawn point tags, but no layers which is the main issue.

![image](https://github.com/user-attachments/assets/226c320d-ca0d-4586-ba80-79df69750c1e)

The spawn point tags used are:

encounter
beacon_bandit
beacon_crawler
beacon_mudraptor

**Comment by Regalis11** [comment]
The layer system issue has now been addressed in our private repo, the changes are coming up in the next unstable build.
